pkg/cloudprovider/providers/gce/gce_instances.go

```diff
@@ -250,50 +253,56 @@ func (gce *GCECloud) AddSSHKeyToAllInstances(user string, keyData []byte) error
 	})
 }
 
-// GetAllZones returns all the zones in which nodes are running
-func (gce *GCECloud) GetAllZones() (sets.String, error) {
-	// Fast-path for non-multizone
-	if len(gce.managedZones) == 1 {
-		return sets.NewString(gce.managedZones...), nil
-	}
-
-	// TODO: Caching, but this is currently only called when we are creating a volume,
-	// which is a relatively infrequent operation, and this is only # zones API calls
+// GetAllCurrentZones returns all the zones in which nodes are currently running
+func (gce *GCECloud) GetAllCurrentZones() (sets.String, error) {
 	zones := sets.NewString()
+	client := gce.client
+	if client == nil {
+		return nil, fmt.Errorf("Client not set on GCECloud object")
+	}
+	nodeList, err := client.Core().Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		log.Fatalf("Failed to list nodes: %v", err)
+		return nil, err
+	}
 
-	// TODO: Parallelize, although O(zones) so not too bad (N <= 3 typically)
-	for _, zone := range gce.managedZones {
-		mc := newInstancesMetricContext("list", zone)
-		// We only retrieve one page in each zone - we only care about existence
-		listCall := gce.service.Instances.List(gce.projectID, zone)
-
-		// No filter: We assume that a zone is either used or unused
-		// We could only consider running nodes (like we do in List above),
-		// but probably if instances are starting we still want to consider them.
-		// I think we should wait until we have a reason to make the
-		// call one way or the other; we generally can't guarantee correct
-		// volume spreading if the set of zones is changing
-		// (and volume spreading is currently only a heuristic).
-		// Long term we want to replace GetAllZones (which primarily supports volume
-		// spreading) with a scheduler policy that is able to see the global state of
-		// volumes and the health of zones.
-
-		// Just a minimal set of fields - we only care about existence
-		listCall = listCall.Fields("items(name)")
-		res, err := listCall.Do()
-		if err != nil {
-			return nil, mc.Observe(err)
-		}
-		mc.Observe(nil)
-
-		if len(res.Items) != 0 {
+	for _, node := range nodeList.Items {
+		labels := node.ObjectMeta.Labels
+		zone, ok := labels[kubeletapis.LabelZoneFailureDomain]
+		if ok {
 			zones.Insert(zone)
+		} else {
+			return nil, fmt.Errorf("Could not find zone for node %v", node.ObjectMeta.Name)
 		}
 	}
-
 	return zones, nil
 }
 
+// GetAllManagedZones gets all the zones managed by k8s
+func (gce *GCECloud) GetAllManagedZones() (sets.String, error) {
```

Comment:
Do we really need this method? I would really like to get rid of it.

Reply:
Currently there is only one place in test using this method and I personally think that we can use GetAllCurrentZones() there too.

Note that currently "managed_zones" can be only set to two values:
- one particular zones
- all zones from the region

There is no way to set it differently in production, thus we probably shouldn't set it differently in tests too.

Reply:
I see your point. I have removed this method and replaced the one use with GetAllCurrentZones